### PR TITLE
Add leakage-weighted distillation loss primitives

### LINKED ISSUE #1010: Add leakage-weighted distillation loss primitives
Parent: #965
Subtask: OM-602.1
Size: M

## Summary
Add the reusable loss primitives needed for leakage-preserving teacher-to-nano distillation. This child covers entity-type weighting, KL on logits, and span-level agreement scoring without introducing the full trainer, mining loop, artifact export, or certification report.

## Scope
- Define deterministic data structures for token logits, labeled spans, entity-type weights, and distillation loss output.
- Compute temperature-scaled KL divergence between teacher and student logits.
- Up-weight loss contributions for leakage-relevant entity types.
- Add span-level boundary agreement loss for teacher/student spans.
- Keep the API dependency-light and offline-friendly for tests.

## Acceptance criteria
- Per-entity weights increase the contribution of rare/leakage-relevant entity types.
- KL loss supports deterministic temperature scaling.
- Span agreement penalizes missed spans and shifted boundaries.
- Tests cover weighting, temperature behavior, deterministic output, and span boundary penalties.

## Out of scope
- Hard-negative mining.
- Feature-matching schedules.
- Certification gates and evidence reports.
- MLX/CoreML artifact conversion.

Part of #965.

### PR DESCRIPTION
**Pull request by maziyarpanahi**
Closes #1010.
Part of #965.

Adds dependency-light distillation primitives for leakage-preserving teacher-to-nano work: entity-type loss weights for hard CE and KL token terms, labeled span agreement loss for missed or shifted boundaries, and focused tests for deterministic weighted output and boundary penalties.

Tests:
- .venv/bin/python -m pytest tests/ -q

### CODE REVIEW

**Comment by maziyarpanahi** [comment]
Reviewed against #1010 and current `master`.

Validation:
- `gh pr checks 1014` passed (15 checks passed/skipped).

This satisfies the linked issue and is ready to merge.
